File: Sources/BNKKit/Resources.swift
import Foundation
import Yams

/// Browsing arbitrary Kubernetes objects.
///
/// Deliberately untyped, unlike the rest of `K8s`: the typed models exist
/// because specific screens render specific fields, and a browser renders
/// whatever is there. So these carry the decoded JSON and pull out only what a
/// list row needs.
/// `@unchecked` because the payload is a JSON dictionary of `Any`, which Swift
/// cannot prove Sendable. It is produced once by `JSONSerialization`, never
/// mutated, and only read — the guarantee holds by construction.
public struct RawObject: Identifiable, @unchecked Sendable {
    public let json: [String: Any]
    public let name: String
    public let namespace: String?
    public let created: Date?

    public var id: String { "\(namespace ?? "-")/\(name)" }

    public init?(_ json: [String: Any]) {
        guard let metadata = json["metadata"] as? [String: Any],
              let name = metadata["name"] as? String else { return nil }
        self.json = json
        self.name = name
        self.namespace = metadata["namespace"] as? String
        self.created = (metadata["creationTimestamp"] as? String)
            .flatMap { KubeClient.rfc3339.date(from: $0) }
    }

    public func string(_ path: String...) -> String? {
        value(path) as? String
    }

    public func int(_ path: String...) -> Int? {
        (value(path) as? NSNumber)?.intValue
    }

    public func array(_ path: String...) -> [[String: Any]] {
        value(path) as? [[String: Any]] ?? []
    }

    private func value(_ path: [String]) -> Any? {
        var current: Any? = json
        for key in path {
            guard let dictionary = current as? [String: Any] else { return nil }
            current = dictionary[key]
        }
        return current
    }

    /// The object as YAML, which is the form anyone reading a spec expects.
    ///
    /// `managedFields` is dropped: it is server bookkeeping, routinely longer
    /// than the object itself, and nobody has ever wanted to read it on a
    /// tablet.
    public var yaml: String {
        var copy = json
        if var metadata = copy["metadata"] as? [String: Any] {
            metadata.removeValue(forKey: "managedFields")
            copy["metadata"] = metadata
        }
        do {
            return try Yams.dump(object: RawObject.plain(copy) ?? [:], sortKeys: true)
        } catch {
            // Say why rather than "could not render", which is what the first
            // version did and taught nobody anything.
            return "could not render this object as YAML: \(error)"
        }
    }

    /// JSONSerialization hands back Foundation's bridged types — `NSNull` for
    /// null and `NSNumber` for every number and boolean — and Yams represents
    /// none of them. Converting to Swift's own types first is what makes the
    /// dump work at all.
    static func plain(_ value: Any) -> Any? {
        switch value {
        case is NSNull:
            return nil
        case let dictionary as [String: Any]:
            return dictionary.reduce(into: [String: Any]()) { out, pair in
                if let converted = plain(pair.value) { out[pair.key] = converted }
            }
        case let array as [Any]:
            return array.compactMap(plain)
        case let number as NSNumber:
            // A JSON true is an NSNumber that happens to be a CFBoolean, and
            // dumping it as 1 would change what the document says.
            if CFGetTypeID(number) == CFBooleanGetTypeID() { return number.boolValue }
            if let exact = number as? Int, Double(exact) == number.doubleValue { return exact }
            return number.doubleValue
        case let string as String:
            return string
        default:
            return String(describing: value)
        }
    }
}

/// A kind the browser can list.
public struct ResourceKind: Identifiable, Hashable, Sendable {
    public let name: String
    public let plural: String
    /// `api/v1` for the core group, `apis/<group>/<version>` for the rest.
    public let root: String
    public let namespaced: Bool

    public var id: String { "\(root)/\(plural)" }

    public func path(namespace: String?) -> String {
        guard namespaced, let namespace, !namespace.isEmpty else { return "/\(root)/\(plural)" }
        return "/\(root)/namespaces/\(namespace)/\(plural)"
    }

    /// What the browser offers.
    ///
    /// Secrets are absent on purpose. Everything else here is safe to put on a
    /// screen; a Secret's whole content is its value, and a generic YAML view of
    /// one would put cluster credentials on a tablet in a coffee shop. The app
    /// reads the two secrets it genuinely needs by name, for certificate dates.
    public static let all: [ResourceKind] = [
        ResourceKind(name: "Pods", plural: "pods", root: "api/v1", namespaced: true),
        ResourceKind(name: "Deployments", plural: "deployments", root: "apis/apps/v1", namespaced: true),
        ResourceKind(name: "DaemonSets", plural: "daemonsets", root: "apis/apps/v1", namespaced: true),
        ResourceKind(name: "StatefulSets", plural: "statefulsets", root: "apis/apps/v1", namespaced: true),
        ResourceKind(name: "Services", plural: "services", root: "api/v1", namespaced: true),
        ResourceKind(name: "ConfigMaps", plural: "configmaps", root: "api/v1", namespaced: true),
        ResourceKind(name: "Events", plural: "events", root: "api/v1", namespaced: true),
        ResourceKind(name: "Nodes", plural: "nodes", root: "api/v1", namespaced: false),
    ]
}

extension KubeClient {
    /// Every object of a kind, as decoded JSON.
    public func list(_ kind: ResourceKind, namespace: String? = nil) async throws -> [RawObject] {
        let data = try await get(kind.path(namespace: namespace))
        let root = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let items = root?["items"] as? [[String: Any]] ?? []
        return items.compactMap(RawObject.init)
    }

    /// The events naming one object, newest first.
    public func events(about name: String, namespace: String) async throws -> [K8s.Event] {
        try await getJSON(K8s.List<K8s.Event>.self, "/api/v1/namespaces/\(namespace)/events",
                          query: ["fieldSelector": "involvedObject.name=\(name)"])
            .items
            .sorted { ($0.at ?? .distantPast) > ($1.at ?? .distantPast) }
    }
}

File: dart/bnk_kit/lib/src/resources.dart
import 'dart:convert';

import 'json.dart';
import 'k8s_types.dart';
import 'kube_client.dart';
import 'time.dart';
import 'yaml_emit.dart';

/// An arbitrary Kubernetes object, as decoded JSON.
///
/// Deliberately untyped, unlike the rest of the models: those exist because
/// specific screens render specific fields, and a browser renders whatever is
/// there. So this carries the decoded JSON and pulls out only what a list row
/// needs.
class RawObject {
  final JsonMap json;
  final String name;
  final String? namespace;
  final DateTime? created;

  RawObject._(this.json, this.name, this.namespace, this.created);

  static RawObject? tryFrom(JsonMap json) {
    final metadata = asMap(json['metadata']);
    final name = metadata?['name'];
    if (metadata == null || name is! String) return null;
    final stamp = metadata['creationTimestamp'];
    return RawObject._(json, name, asString(metadata['namespace']),
        stamp is String ? Rfc3339.parse(stamp) : null);
  }

  String get id => '${namespace ?? '-'}/$name';

  String? string(List<String> path) => asString(_value(path));

  int? integer(List<String> path) => asInt(_value(path));

  List<JsonMap> array(List<String> path) => asList(_value(path), (m) => m);

  Object? _value(List<String> path) {
    Object? current = json;
    for (final key in path) {
      if (current is! Map) return null;
      current = current[key];
    }
    return current;
  }

  /// The object as YAML, which is the form anyone reading a spec expects.
  ///
  /// `managedFields` is dropped: it is server bookkeeping, routinely longer
  /// than the object itself, and nobody has ever wanted to read it on a
  /// tablet. Nulls are dropped too, as a YAML dump of the object would.
  String get yaml {
    final copy = _withoutNulls(json) as Map;
    final metadata = copy['metadata'];
    if (metadata is Map) metadata.remove('managedFields');
    return emitYaml(copy, sortKeys: true);
  }

  static Object? _withoutNulls(Object? value) {
    if (value is Map) {
      return <String, dynamic>{
        for (final e in value.entries)
          if (e.value != null) e.key.toString(): _withoutNulls(e.value)
      };
    }
    if (value is List) {
      return [for (final e in value) if (e != null) _withoutNulls(e)];
    }
    return value;
  }
}

/// A kind the browser can list.
class ResourceKind {
  final String name;
  final String plural;

  /// `api/v1` for the core group, `apis/<group>/<version>` for the rest.
  final String root;
  final bool namespaced;

  const ResourceKind(
      {required this.name,
      required this.plural,
      required this.root,
      required this.namespaced});

  String get id => '$root/$plural';

  String path(String? namespace) {
    if (!namespaced || namespace == null || namespace.isEmpty) {
      return '/$root/$plural';
    }
    return '/$root/namespaces/$namespace/$plural';
  }

  /// What the browser offers.
  ///
  /// Secrets are absent on purpose. Everything else here is safe to put on a
  /// screen; a Secret's whole content is its value, and a generic YAML view of
  /// one would put cluster credentials on a tablet in a coffee shop. The app
  /// reads the two secrets it genuinely needs by name, for certificate dates.
  /// Nodes lead deliberately: the first entry is what the screen opens on.
  static const all = [
    ResourceKind(name: 'Nodes', plural: 'nodes', root: 'api/v1', namespaced: false),
    ResourceKind(name: 'Pods', plural: 'pods', root: 'api/v1', namespaced: true),
    ResourceKind(name: 'Deployments', plural: 'deployments', root: 'apis/apps/v1', namespaced: true),
    ResourceKind(name: 'DaemonSets', plural: 'daemonsets', root: 'apis/apps/v1', namespaced: true),
    ResourceKind(name: 'StatefulSets', plural: 'statefulsets', root: 'apis/apps/v1', namespaced: true),
    ResourceKind(name: 'Services', plural: 'services', root: 'api/v1', namespaced: true),
    ResourceKind(name: 'ConfigMaps', plural: 'configmaps', root: 'api/v1', namespaced: true),
    ResourceKind(name: 'Events', plural: 'events', root: 'api/v1', namespaced: true),
  ];

  @override
  bool operator ==(Object other) => other is ResourceKind && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

extension ResourcesApi on KubeClient {
  /// Every object of a kind, as decoded JSON.
  Future<List<RawObject>> list(ResourceKind kind, {String? namespace}) async {
    final data = await get(kind.path(namespace));
    final root = jsonDecode(utf8.decode(data));
    final items = root is Map ? root['items'] : null;
    return [
      for (final item in asList(items, (m) => m))
        if (RawObject.tryFrom(item) case final object?) object
    ];
  }

  /// The events naming one object, newest first.
  Future<List<Event>> events(
      {required String about, required String namespace}) async {
    final items = await getJson('/api/v1/namespaces/$namespace/events',
        (j) => listItems(j, Event.fromJson),
        query: {'fieldSelector': 'involvedObject.name=$about'});
    final epoch = DateTime.fromMillisecondsSinceEpoch(0);
    items.sort((a, b) => (b.at ?? epoch).compareTo(a.at ?? epoch));
    return items;
  }
}
